src/relations/osm_mysql.py

```diff
@@ -0,0 +1,190 @@
+
+        return password
+
+    def _on_leader_elected(self, _) -> None:
+        """Handle the leader elected event.
+
+        Retrieves relation data from the peer relation databag and copies
+        the relation data into the new leader unit's databag.
+        """
+        # Skip if the charm is not past the setup phase (config-changed event not executed yet)
+        if not self.charm._is_peer_data_set:
+            return
+
+        relation_data = json.loads(
+            self.charm._peers.data[self.charm.app].get("osm_mysql_relation_data", "{}")
+        )
+
+        for relation in self.charm.model.relations.get(LEGACY_OSM_MYSQL, []):
+            relation_databag = relation.data
+
+            # Copy relation data into the new leader unit's databag
+            for key, value in relation_data.items():
+                if relation_databag[self.charm.unit].get(key) != value:
+                    relation_databag[self.charm.unit][key] = value
+
+            # Assign the cluster primary's address as the database host
+            primary_address = self.charm._mysql.get_cluster_primary_address()
+            if not primary_address:
+                self.charm.unit.status = BlockedStatus(
+                    "Failed to retrieve cluster primary address"
+                )
+                return
+
+            relation_databag[self.charm.unit]["host"] = primary_address.split(":")[0]
+
+    def _on_osm_mysql_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handle the legacy 'mysql' relation created event.
+
+        Will set up the database and the scoped application user. The connection
+        data (relation data) is then copied into the peer relation databag (to
+        be copied over to the new leader unit's databag in case of a new leader
+        being elected).
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        # Wait until on-config-changed event is executed
+        # (wait for root password to have been set)
+        if not self.charm._is_peer_data_set:
+            event.defer()
+            return
+
+        logger.warning("DEPRECATION WARNING - `osm-mysql` is a legacy interface")
+
+        username = self.charm.config.get("osm-mysql-interface-user")
+        database = self.charm.config.get("osm-mysql-interface-database")
+
+        # Ensure that config values exist, else we will be unable to create database and user
+        if not username or not database:
+            self.unit.status = BlockedStatus("Missing `osm-mysql` relation data")
+            return
+
+        user_exists = False
+        try:
+            user_exists = self.charm._mysql.does_mysql_user_exist(username, "%")
+        except MySQLCheckUserExistenceError:
+            self.charm.unit.status = BlockedStatus("Failed to check user existence")
+            return
+
+        # Only execute if the application user does not exist
+        if user_exists:
```

Comment:
If i understood correctly, this case will be valid when the joining unit is from a second application being related using `osm-mysql`.
right? or am I out to lunch?

Reply:
you're correct. we cannot prevent multiple applications from being related on the same interface/endpoint. this check is just address this edge case